src/v/cloud_topics/level_zero/stm/ctp_stm.cc

```diff
@@ -112,8 +169,37 @@ ss::future<> ctp_stm::apply_raft_snapshot(const iobuf& buf) {
 }
 
 ss::future<iobuf> ctp_stm::take_raft_snapshot(model::offset snapshot_at) {
-    auto st = _state.get_state_at(snapshot_at);
-    co_return serde::to_iobuf(std::move(st));
+    vassert(
+      last_applied() >= snapshot_at,
+      "The snapshot is taken at offset {} but current insync offset is {}",
+      snapshot_at,
+      last_applied());
+    co_return serde::to_iobuf(_state);
+}
+
+ss::future<cluster_epoch_fence> ctp_stm::fence_epoch(cluster_epoch e) {
+    auto term = _raft->confirmed_term();
+    auto projected_epoch = _state.get_projected_epoch();
+    if (projected_epoch.has_value() && projected_epoch.value() == e) {
+        // Case 1. Same epoch, need to acquire read-lock.
+        auto unit = co_await _lock.hold_read_lock();
+        // Invariant: the projected epoch is not nullopt because once
+        // set the projected epoch is never resets.
+        if (_state.get_projected_epoch() == e) {
+            // The projected epoch didn't advance after the scheduling point
+            co_return cluster_epoch_fence{std::move(unit), term};
+        }
+    } else {
+        // Case 2. New epoch, need to acquire write-lock.
+        auto unit = co_await _lock.hold_write_lock();
+        auto current_epoch = _state.get_projected_epoch();
+        if (!current_epoch.has_value() || current_epoch.value() <= e) {
+            _state.advance_projected_epoch(e);
+            co_return cluster_epoch_fence{std::move(unit), term};
+        }
```

Comment:
hmm, it's surprising to me that we have any concurrency to extract within the state machine update path that would benefit from a read/write lock over a mutex... but will keep my eye out